src/json/from_json.cr

```diff
@@ -391,26 +391,38 @@ struct Time::Format
 end
 
 module JSON::ArrayConverter(Converter)
-  def self.from_json(pull : JSON::PullParser)
-    ary = Array(typeof(Converter.from_json(pull))).new
-    pull.read_array do
-      ary << Converter.from_json(pull)
+  private struct WithInstance(T)
+    def from_json(pull : JSON::PullParser)
+      ary = Array(typeof(@converter.from_json(pull))).new
+      pull.read_array do
+        ary << @converter.from_json(pull)
+      end
+      ary
     end
-    ary
+  end
+
+  def self.from_json(pull : JSON::PullParser)
+    WithInstance.new(Converter).from_json(pull)
   end
 end
 
 module JSON::HashValueConverter(Converter)
-  def self.from_json(pull : JSON::PullParser)
-    hash = Hash(String, typeof(Converter.from_json(pull))).new
-    pull.read_object do |key, key_location|
-      parsed_key = String.from_json_object_key?(key)
-      unless parsed_key
-        raise JSON::ParseException.new("Can't convert #{key.inspect} into String", *key_location)
+  private struct WithInstance(T)
```

Comment:
Minor, but I don't see a reason for the `With` prefix.

```suggestion
  private struct Instance(T)
```

Reply:
IMO `Instance` is confusing even if only used internally, because that has connotations of the singleton pattern.